### PR TITLE
1. Added motion vector as 2nd output from the H264EncoderV4l2 module.

### LINKED ISSUE #240: Add motion vector frame as 2nd output pin from H264EncoderV4L2 module
**Is your feature request related to a problem? Please describe.**
To implement 2nd output from H264EncoderV4L2 module as motion vector .

**Describe the solution you'd like**
During h264 encoding motion vectors are generated as difference in motion vectors between two subsequent frames. It is the process by which frames can be assessed for similarities between frames that can reused. The same vectors can be generated as by product while h264 encoding .

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by mohammedzakikochargi**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #240 

**Description**
Implemented 2nd output from H264EncoderV4L2 module as motion vector frame apart from h264 encoded frames

Precise description of the changes in this pull request
During h264 encoding motion vectors are generated as difference in motion vectors between two subsequent frames. It is the process by which frames can be assessed for similarities between frames that can reused. The same motion vectors are  extracted and serialized using overlay module, and it is send as 2nd output from H264EncoderV4L2 module apart from h264 encoded frames. Also added props to enable or disable motion vector frame

**Alternative(s) considered**

Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**

Type Choose one: (Feature)

**Screenshots (if applicable)**

**Checklist**

- [x] I have read the [Contribution Guidelines](https://github.com/Apra-Labs/ApraPipes/wiki/Contribution-Guidelines)
- [x] I have written Unit Tests
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach

### CODE REVIEW

File: base/src/H264EncoderV4L2.cpp [hunk 1]
```diff
@@ -5,12 +5,15 @@
...
 #include "Logger.h"
 #include "Utils.h"
+#include "H264Metadata.h"
 #include "AIPExceptions.h"
 
 H264EncoderV4L2::H264EncoderV4L2(H264EncoderV4L2Props props) : Module(TRANSFORM, "H264EncoderV4L2", props), mProps(props)
 {
-	mOutputMetadata = framemetadata_sp(new FrameMetadata(FrameMetadata::H264_DATA));
-	mOutputPinId = addOutputPin(mOutputMetadata);
+	mOutputMetadata = framemetadata_sp(new H264Metadata(0, 0));
```
Comment (mraduldubey) [inline base/src/H264EncoderV4L2.cpp:13]: link the nvidia sample used to refer to this

File: base/src/H264EncoderV4L2Helper.cpp [hunk 2]
```diff
@@ -200,10 +236,100 @@ int H264EncoderV4L2Helper::setExtControls(v4l2_ext_control &control)
...
+    auto motionVectorFrame = makeFrame(1024 * 1024 * 3, motionVectorFramePinId);
+    uint32_t *frameBuffer = reinterpret_cast<uint32_t *>(motionVectorFrame->data());
+    size_t mCount = 0;
+    for (uint32_t i = 0; i < numMVs; i++, pInfo++)
+    {
+
+        if (abs(pInfo->mv_x) > motionVectorThreshold || abs(pInfo->mv_y) > motionVectorThreshold)
+        {
+            auto tempY = floor(i / totalMacroblockInRow);
+            auto y = tempY * 16 + 8;
```
Comment (mraduldubey) [inline base/src/H264EncoderV4L2Helper.cpp:288]: 16 and 8 are harcoded values - why ? at least, put comments explaining these

File: base/src/H264EncoderV4L2Helper.cpp [hunk 3]
```diff
@@ -200,10 +236,100 @@ int H264EncoderV4L2Helper::setExtControls(v4l2_ext_control &control)
...
+            auto tempX = floor(i % totalMacroblockInRow);
+            auto x = tempX * 16 + 8;
+            CircleOverlay circleOverlay;
+            circleOverlay.x1 = x;
+            circleOverlay.y1 = y;
+            circleOverlay.radius = 1;
+            frameBuffer[mCount] = static_cast<uint32_t>(x);
```
Comment (mraduldubey) [inline base/src/H264EncoderV4L2Helper.cpp:295]: these three lines are not required

File: base/test/h264encoderv4l2_tests.cpp [hunk 4]
```diff
@@ -71,13 +76,13 @@ BOOST_AUTO_TEST_CASE(rgb24_1280x720, *boost::unit_test::disabled())
 	auto encoder = boost::shared_ptr<Module>(new H264EncoderV4L2(encoderProps));
 	copy->setNext(encoder);
 
-	auto fileWriter = boost::shared_ptr<Module>(new FileWriterModule(FileWriterModuleProps("./data/testOutput/Raw_RGB24_1280x720.h264", true)));
-	encoder->setNext(fileWriter);
+	auto sink = boost::shared_ptr<ExternalSinkModule>(new ExternalSinkModule());
+	encoder->setNext(sink);
```
Comment (mraduldubey) [inline base/test/h264encoderv4l2_tests.cpp:80]: why cudamemcpy ? - remove this test if not useful

File: base/test/h264encoderv4l2_tests.cpp [hunk 5]
```diff
@@ -212,4 +218,114 @@ BOOST_AUTO_TEST_CASE(encodepush, *boost::unit_test::disabled())
...
+	BOOST_TEST(sink->init());
+
+	fileReader->play(true);
+
+	for (auto i = 0; i < 40; i++)
+	{
+		fileReader->step();
+		encoder->step();
+		auto frames = sink->pop();
+		for (auto it = frames.cbegin(); it != frames.cend(); it++)
```
Comment (mraduldubey) [inline base/test/h264encoderv4l2_tests.cpp:251]: add a check for count of frames for which you got motion data - outside of the for loop

**Review by mraduldubey — CHANGES_REQUESTED** [review]
remove tests which are not required

File: base/include/H264EncoderV4L2Helper.h [hunk 6]
```diff
@@ -1,19 +1,18 @@
 #pragma once
+#include <boost/pool/object_pool.hpp>
 
+#include "ExtFrame.h"
 #include "AV4L2ElementPlane.h"
 #include "V4L2CUYUV420Converter.h"
-
-#include "ExtFrame.h"
-#include <boost/pool/object_pool.hpp>
+#include "v4l2_nv_extensions.h"
```
[the remark this answers is not in the document]
Comment (mohammedzakikochargi) [reply base/include/H264EncoderV4L2Helper.h:7]: Done

File: base/src/H264EncoderV4L2.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (mohammedzakikochargi) [reply base/src/H264EncoderV4L2.cpp:13]: Made it as a temp variable

File: base/src/H264EncoderV4L2.cpp [hunk 7]
```diff
@@ -144,12 +142,18 @@ bool H264EncoderV4L2::processSOS(frame_sp &frame)
 		v4l2MemType = V4L2_MEMORY_DMABUF;
 	}
 
-	mHelper = H264EncoderV4L2Helper::create(v4l2MemType, pixelFormat, width, height, step, 1024 * mProps.targetKbps, 30, [&](frame_sp &frame) -> void {
-		frame->setMetadata(mOutputMetadata);
-
-		frame_container frames;
-		frames.insert(make_pair(mOutputPinId, frame));
-		send(frames);
+	auto h264OutMetadata = framemetadata_sp(new H264Metadata(width, height));
```
[the remark this answers is not in the document]
Comment (mohammedzakikochargi) [reply base/src/H264EncoderV4L2.cpp:145]: Removed the down casting

File: base/src/H264EncoderV4L2Helper.cpp (hunk 2 above)
Comment (mohammedzakikochargi) [reply base/src/H264EncoderV4L2Helper.cpp:288]: Added comments explaining about the values

File: base/src/H264EncoderV4L2Helper.cpp (hunk 3 above)
Comment (mohammedzakikochargi) [reply base/src/H264EncoderV4L2Helper.cpp:295]: Removed the three lines

File: base/test/h264encoderv4l2_tests.cpp (hunk 4 above)
Comment (mohammedzakikochargi) [reply base/test/h264encoderv4l2_tests.cpp:80]: This module accepts both host and cuda memory

File: base/test/h264encoderv4l2_tests.cpp (hunk 5 above)
Comment (mohammedzakikochargi) [reply base/test/h264encoderv4l2_tests.cpp:252]: done

**Review by mraduldubey — APPROVED** [review]

**Pull request merged**